FIP-0086 updates from implementation (#1022)
filecoin-project/FIPs: FIPS/fip-0086.md

### F3 Overview and Interaction with EC

This proposal introduces an F3 component, which works alongside EC in Filecoin client nodes (participants).

The participants in F3 are the storage providers (SPs) of the Filecoin network. The participation of each SP is weighted according to its quality-adjusted power (QAP), which is a function of the storage power that the SP has committed to the network. This information is maintained in a built-in actor called the _power actor_ (f04). Only SPs that hold more than the threshold of power to participate in EC can participate in F3.

In short, each participant $p$ in the Filecoin network (i.e., a storage provider with power) runs F3 in a loop and feeds its input from EC. F3 returns a finalized prefix chain to EC, which updates the canonical chain to extend this F3-finalized prefix. More precisely, in each loop iteration $i$ (corresponding to the i-th instance of F3):

> - **EC/F3 interface:** Participant $p$ feeds its current canonical chain $canonical$ and its previously F3-finalized chain, called the $baseChain$, to F3. The $baseChain$ (with some lookback parameter) fixes the power table used by F3 and randomness seed used to configure the F3 instance, while $p$'s current canonical chain is the chain $p$ proposes to be finalized.

Reviewer: maybe worth specifying what is the loockback parameter?

Reviewer: (later on line 111 it specifies the value to be 10 it seems like

### F3 Adversarial Model

Honest participants follow the protocol at all times. Byzantine participants deviate arbitrarily from the protocol. Rational participants deviate from the protocol to pursue a greater (expected) pay-off according to their utility function. A Byzantine adversary is able to control and orchestrate up to less than ⅓ QAP.

The actions of honest and rational participants are aligned by ensuring that rational participants never contribute to finalizing a chain they do not locally possess. Without this, If a participant $p$ were to finalize a chain it didn't have, then $p$ would not be able to mine in EC until it retrieved the required blocks and thus, in the meantime, could not obtain the cryptoeconomic rewards that EC offers. To this end, F3 provides _EC compatibility_.

F3 and its implementation GossiPBFT provide additional robustness for strong adversaries that control more than ⅓ QAP, such as:

* Censorship resistance against a coalition of rational participants trying to leak power, without providing firm guarantees.
* Resilience to long-range attacks by an adversary controlling up to less than ⅔ QAP of any old power table.

> ### Best-effort Broadcast and GossipSub

Author: I'll change all to Gossipsub to match that project's naming.

F3 uses GossipSub (like Filecoin EC) to disseminate protocol messages, implementing a broadcast channel among participants. GossipSub is assumed to satisfy the following properties of the _best-effort broadcast primitive_ (denoted _BEBroadcast_ hereafter):

* _No creation_: No message with sender $p$ is delivered unless it was previously broadcast by $p$.
* _Validity_: If $p$ and $p'$ are honest, then every message broadcast by $p$ is eventually delivered by $p'$.

> Note that academic nomenclature may be misleading, and "every message broadcast is eventually delivered" does not mean the broadcast is reliable. Messages may be lost if the sender, receiver, or link fail, but will be delivered if they don't.

Reviewer 2: I don't think this is a matter of misleading language. This is merely a stated assumption, which (as is often the case) doesn't 100% match reality (e.g. if honest $p'$ dies and doesn't recover, it obviously won't receive it).

Author: I've reworded this to make explicit the assumptions behind "honest".

The _no duplication_ property of a traditional best-effort broadcast primitive is not needed; indeed F3 explicitly rebroadcasts messages to overcome loss. 

### Partially Synchronous Model and Δ-Synchrony

A system is $\Delta$-_synchronous_ if the combined computation and communication delay between any two honest participants is smaller than $\Delta$. Under $\Delta$-synchrony, any message broadcast by an honest participant is delivered by every honest participant within a time bound of $\Delta$.

> In practice, if GossipSub is used for _BEBroadcast_, $\Delta$ is an assumed upper bound on GossipSub latency (e.g., a few seconds).

Reviewer 2: ```suggestion
In practice, if GossipSub is used for _BEBroadcast_, $\Delta$ is an assumed upper bound on GossipSub latency.
```

Doesn't add much information.

For termination, a classical partially synchronous model is assumed, with an unknown bound on communication delays among non-Byzantine participants.

### Power Table and Consensus Committee

The EC chain state maintains a table of the quality-adjusted power of each SP, as well as their signing key. Each tipset $t$ in the chain commits to a particular state that results from executing all messages in the chain from genesis to $t$. The committee of participants for an instance of GossiPBFT is determined by the chain state resulting from the tipset finalized by a previous instance: the input parameter $committee$ of instance $i$ is defined by the tipset output by instance $i-PowerTableLookback$. If $i$ is less than $PowerTableLookback$, $committee$ is drawn from the chain state committed by the genesis.

Given an EC chain state at some epoch, the corresponding F3 committee comprises all miner actors with:
- at least 10TiB of quality-adjusted power;
- zero fee debt; and

> - no active consensus fault.

Author: @Kubuxu could you please check the EC leader election and confirm there are no other relevant checks? I'm aware that in EC these are applied against both lookback state and current state: we can only do one.

Reviewer 3: In the code, I'm using the same rules that apply to the EC election but without the Finality lookback.

Author: That sounds great. Can you please clarify that I have captured those rules here? Are there any more?

Reviewer 3: There is a caveat about not having enough miners (if there are fewer miners than X, the minimum power no longer applies). Other than that, it is the 3 checks you mentioned.

Reviewer 4: Should we add that line here that the number of miners must be X to reach consensus?

Author: > Should we add that line here that the number of miners must be X to reach consensus?

There's no minimum on the number of participants, except implicitly 1. The protocol would be unable to recover from dropping below any such number.

The F3 consensus power of each eligible F3 participant is the corresponding miner actor's EC quality-adjusted power, scaled such that the total power of all participants fits in a 16-bit unsigned integer.

```

> ParticipantPower ← truncate(0xffff * MinerQAP / TotalQAP)

Author: @Stebalien I inferred this from code. But currently this is an over-simplification because the code also preserves unscaled values, and uses them in ticket computations. 

Can we make it an accurate simplification by removing unscaled power from the code?

Reviewer 5: Probably? Although I'm not sure what the impact of a ticket clash is (I assume rather small).

Reviewer 5: Fixed.

```

The total F3 consensus weight is the sum of all eligible participants' consensus weight (note this differs from EC leader election). An F3 participant's signing key is the associated miner actor's worker key (always a BLS key).

The parameter $PowerTableLookback$ takes the value 10.

### Properties of F3

> F3 is not identical to classical consensus, however similar to it. In a nutshell, the reasons why EC and Filecoin should not rely on classical consensus properties (implemented by existing off-the-shelf BFT consensus protocol) for fast finality are twofold: EC-compatibility and resilience to Filecoin power leakage attacks in case of a strong adversary. 

Reviewer 2: ```suggestion
F3 is not identical to classical consensus, however similar to it. In a nutshell, the reasons why EC and Filecoin should not rely on classical consensus properties (implemented by existing off-the-shelf BFT consensus protocol) for fast finality are twofold: EC compatibility and resilience to Filecoin power leakage attacks in case of a strong adversary. 
```

It's actually defined as EC compatibility (no hyphen) above.

With this in mind, the F3 component interface and properties are given below.

> **Interface and properties of F3 at participant p:**
>

> > `F3.invoke (int i, chain canonical, chain baseChain)` **returns** ("finalizes") `(int i, chain h, proof_of_finality certificate)`

Reviewer 2: Following the changes above, a "proof of finality" is never introduced or defined until line 129 below, and even then it's a bit shaky. Certificate is also not defined as a value of type proof_of_finality. Could it be more intuitive to just use `finality_certificate fc` or something alike?

>
> **Properties:**
>
> **Agreement.** If two honest participants finalize $(i,h,\ast)$ and $(i,h',\ast)$, then $h = h'$.
>
> **Validity.** If an honest participant finalizes $(i,h,\ast)$, then $h$ is a prefix of the canonical input chain of some honest participant $p'$ in instance $i$.
>

> > **Proof of Finality.** If an honest participant finalizes $(i,h,certificate)$, then certificate is signed by ⅔ QAP majority corresponding to the committee input to instance $i$ of some honest participant $p'$.

Reviewer 2: ```suggestion
> **Proof of Finality.** If an honest participant finalizes $(i,h,certificate)$, then $certificate$ is signed by ⅔ QAP majority corresponding to the committee input to instance $i$ of some honest participant $p'$.
```

But see previous comment.

```
Verify(certificate, decision, committee) → boolean
```

This function uses the certificate to verify that $decision$ was the output of some instance of consensus executed among the $committee$.

### F3 Pseudocode

The F3 algorithm is specified by pseudo-code:

```
// A list of all finalized tipsets with their respective certificates.
// One entry per consensus instance.
// Initialized with a genesis which needs no proof.
finalizedTipsets[0] ← {tipset: genesis, certificate: nil}

// The first instance number after genesis is 1.
i ← 1
while True:
   // Delay start until EC is 2 epochs ahead of last finalised chain.

>   wait until EC.CurrentEpoch() >= finalizedTipsets[i-1].Head().epoch + 2

Reviewer 3: We can't do that exactly. We want strict timing on instance start, so it should depend only on the previously finalized chain. This depends on each participant's local view of EC chain.

Now I noticed that it is already part of the spec.

Author: `EC.CurrentEpoch()` is a bit ambiguous here, but I interpreted it to mean "when the clock says the epoch has started", with no reference to the chain. Does that match your understanding?

Reviewer 3: Yeah, that makes sense, I read the spec as `loop on this check until ChainHead.Epoch() >= ...` but I'm fine moving on from this.

>
> **EC compatibility.** An honest participant never contributes to gathering ⅔ QAP (super-majority) of votes for a chain $c$ which its local EC instance did not already deliver unless it already observes evidence of such a super-majority for chain $c$.

> The $\Delta$ parameter is specified (See [below](#Synchronization-of-Participants-in-the-Current-Instance)).

Reviewer 2: ```suggestion
The $\Delta$ parameter is specified [below](#Synchronization-of-Participants-in-the-Current-Instance).
```

This was extremely dangling. It's still a bit dangling. Do we need this line?

```
Verify(certificate, decision, committee) → boolean
```

This function uses the certificate to verify that $decision$ was the output of some instance of consensus executed among the $committee$.

### F3 Pseudocode

The F3 algorithm is specified by pseudo-code:

```
// A list of all finalized tipsets with their respective certificates.
// One entry per consensus instance.
// Initialized with a genesis which needs no proof.
finalizedTipsets[0] ← {tipset: genesis, certificate: nil}

// The number of epochs by which instance start is delayed relative to 

> // epoch of last finalised chain head.

Reviewer 6: The FIP has a mix of British-Engish and American-english wording. We probably want to do a scan of all such cases and pick one (probably American-english)? Example: the text contains both "finalise" and "finalize" in number of places.

Reviewer 2: I'd rather go the other way, but that's probably the sensible choice. But let's definitely make it consistent.

Author: Changed to "finalize", and there were no more "ise" that aren't consistent between dialects.

i ← 1
while True:
  // Delay start until EC is ahead of last finalised chain plus instance start backoff.
  wait until EC.CurrentEpoch() >= finalisedTipsets[i-1].Head().epoch + startBackoffEpochs
  // Delay start until EC is 2 epochs ahead of last finalised chain.
  wait until EC.CurrentEpoch() >= finalizedTipsets[i-1].Head().epoch + 2 
  // Fetch tipsets from locally-observed EC chain from (inclusive) the last finalised tipset.
  proposal ← EC.HeaviestUnfinalizedChain()
  if proposal.Head().epoch = EC.CurrentEpoch()
    // Remove current epoch tipset, as pre-agreement is unlikely
    proposal ← proposal[:-1]
  if proposal.Head() = finalizedTipsets[i-1].Head()
    // EC has produced empty tipsets. Increase backoff parameters, and retry.
    startBackoffEpochs ← startBackoffEpochs + nextStartBackoffIncrement(startRetryAttempts)
    startRetryAttempts ← startRetryAttempts + 1
    continue // Skip to the next iteration.
  else
    // Reset backoff parameters. 
    startBackoffEpochs ← 0
    startRetryAttempts ← 0
  // Get the power table from previously-finalised tipset from lookback instances ago.
  // Use genesis power for initial instances.

>   lookback ← min(i, PowerTableLookback)

Reviewer 2: Is `PowerTableLookback` supposed to be EC's 900? It's undefined here.

Author: It was defined at 111 (= `10`).

The function `nextStartBackoffIncrement(int)` returns the number of epochs by which `startBackoffEpochs` should be increased based on the number of attempts, referred to as `startRetryAttempts`. Implementers may choose any increment function at their discretion, but for each instance number `i` the increment amount relative to the number of attempts must be consistent across all participants.

### Changes to EC: Fork Choice Rule

The current EC fork-choice rule selects, from all the known tipsets, the tipset backed by the most weight.
This fork choice rule is modified to accommodate the finalization of tipsets by F3. **This is the only change to EC this FIP introduces.**

The fork choice rule is modified in the presence of a finalized prefix:
* the fork choice rule selects the heaviest chain out of all chains with a prefix that **includes all tipsets finalized by F3**.

A competing tipset $t'$ that is a superset of finalized tipset $t$ at the same epoch is not heavier than $t$ itself, despite it being backed by more EC power.

> This redefinition of the fork choice rule is consistent with the abstract notion of the heaviest chain being backed by the most power because a finalized tipset has been backed by a super-majority of the committee in GossiPBFT. In contrast, any non-finalized block in the same epoch is only backed in that epoch by it's proposer.

Reviewer 2: ```suggestion
This redefinition of the fork choice rule is consistent with the abstract notion of the heaviest chain being backed by the most power because a finalized tipset has been backed by a super-majority of the committee in GossiPBFT. In contrast, any non-finalized block in the same epoch is only backed in that epoch by its proposer.
```

The algorithm starts by initializing $finalizedTipsets[0]$ with the genesis tipset. This tipset is pre-defined as finalized and does not require a PoF. Then, in each iteration $i$ of the loop, it takes the following steps:

1. Wait until the current epoch is at least two epochs from that of the latest finalized head.
2. Obtain the set of participants of instance $i$ from the power table determined by the finalized tipset $PowerTableLookback$ instances ago.
3. Call the $\texttt{HeaviestUnfinalizedChain}()$ function that returns the tipsets of the locally observed chain constructed by EC from finalizedTipset[i-1].Head (including the head of the finalized chain) and sets the $proposal$ variable to the returned value. Note that the proposal is pruned so that its head is at most a tipset from the previous epoch, not the current epoch.
4. Execute the instance $i$ of GossiPBFT consensus.
5. Add the returned tipset from the consensus execution to the $finalizedTipsets$ list.

> Note that if EC produces empty tipsets, EC epochs progress regardless. F3 will then repeatedly propose and finalize the same base chain until a non-empty tipset is produced. 

Author: This is a semantic change, there is no such blocking implemented. It seems difficult to add because participants would observe the first interesting chain at different times and so become unsynchronised. We need them all to start at the ~same time. 

There could be a more complicated function of how many times the same tipset has been finalised and its epoch to compute an increasing delay.

Reviewer 5: cc @Kubuxu you had a proposal, IIRC?

Reviewer 3: The proposal I had was something akin to back-off when we finalize base. The time gets larger (up to a cap) each time we observe the base is finalized. It would depend on the previous instances.

Reviewer 6: This proposal is now implemented/merged.

Author: Please either push a commit to or make a PR on this PR

### GossiPBFT Consensus

This section provides the specification of GossiPBFT, the consensus protocol that is iteratively instantiated by the F3 loop and returns a decision (in the form of an F3-finalized chain) and a certificate per instance.

> GossiPBFT is a Byzantine fault-tolerant consensus protocol that is resilient optimal, i.e., it tolerates up to less than ⅓ QAP being controlled by a Byzantine adversary. The committee for each instance of the protocol is known. Each participant inputs a proposal value, and the protocol outputs one of the input values as the final decision value. Unlike a longest-chain protocol, the output of GossiPBFT is permanent.

Reviewer 2: ```suggestion
GossiPBFT is a Byzantine fault-tolerant consensus protocol that is resilient-optimal, i.e., it tolerates up to less than ⅓ QAP being controlled by a Byzantine adversary. The committee for each instance of the protocol is known. Each participant inputs a proposal value, and the protocol outputs one of the input values as the final decision value. Unlike a longest-chain protocol, the output of GossiPBFT is permanent.
```

This should likely be _resilient-optimal_ (i.e. with hyphen) or _optimally resilient_. It doesn't make much sense as is, but perhaps it's accepted in the field.

}
```

Values for step numbers are:

| Step name | Value |
|-----------|-------|
| QUALITY   | 1     |
| CONVERGE  | 2     |
| PREPARE   | 3     |
| COMMIT    | 4     |
| DECIDE    | 5     |

All messages broadcast by a participant have their participant ID in the sender field and contain a digital signature by that participant $(m.Signature)$ over `"GPBFT:filecoin:":15 || Step:1 || Round:8 || Instance:8 || SupplementalData.Commitments:32 || MerkelizeValue(Value):32 || SUpplementalData.PowerTable` (where `Round` and `Instance` are big-endian encoded). The protocol assumes aggregatable signatures (e.g., BLS, Schnorr), resilient to [rogue public key attacks](https://crypto.stanford.edu/~dabo/pubs/papers/BLSmultisig.html) (see [Boneh, Drijvers, and Neven](https://eprint.iacr.org/2018/483.pdf) construction).

The receiver of a message only considers messages with matching supplemental data and valid signatures, and discards all other messages as invalid. The sender ID and signatures are omitted in further descriptions for better readability.

> Two (or more) messages $m1$ and $m2$ are called _equivocating messages_ if $m1.Sender=m2.Sender \land m1.Instance=m2.Instance \land m1.Value ≠ m2.Value \land m1.MsgType=m2.MsgType \land \texttt(if applicable)\ m1.Round=m2.Round$. $m1.Sender$ is called an _equivocating sender_.

Author: Replaced with prose.

57:  return (StrongQuorumValue(M), Aggregate(M)) // Terminate with a decision
```

A participant jumps immediately to the DECIDE step upon receiving a DECIDE message from another participant. That message will carry evidence of a strong quorum of the committee COMMITting to that value.

```
     // Decide anytime
58:  upon reception of a valid <DECIDE, v, evidence>, if step != DECIDE
59:    reBroadcast <DECIDE, v, evidence>
60:    go to line 56
```

The helper function mayHaveStrongQuorum returns False if, given the already delivered messages, the participant knows for a fact that no correct participant could have decided the given value in the given round, even in the presence of an adversary controlling advPower the relative amount of total QAP equivocating, and True otherwise. The parameter advPower can be set to 0 in order to consider the possibility of the participant locally deciding, whereas when set to 1/3 it considers the possibility of any participant deciding:

```
61:  mayHaveStrongQuorum(value, round, step, advPower): 
62:    M  ← { m | m.step = step AND m.round = round AND m is valid} // Clean set of messages
63:    M' ← { m | m ∈ M AND m.value = value } 

> 64:    return Power(M') + (1-Power(M)) + advPower > ⅔ :  

Author: @ranchalp I believe this expression is more clear. However there's an edge condition change here in the change of operator from `<` to `>`. I think the code I've proposed here does what it says: if all the power adds up only to _equal to 2/3_ then there is no strong quorum possible. But I recall this edge case came up in earlier discussions and I was unable to recall if it was critical.

I can change the operator to `>=`, but then I think we'd want a clear description of why the `=` case was considered a possible strong quorum.

Reviewer 7: You are right, it should be `>2/3`and not `>=2/3`. However I noticed the FIP defines a strong quorum as `Power(p) >= \lceil Power(P)*2/3 \rceil` so that should be addressed and changed to `>2/3`. This is only ever a problem if the total power is divisible by 3 and there is a partition in 3 equal parts, one of them Byzantine.

Reviewer 7: Actually, since the adversary always hold strictly less than a third of the total power, the definition of strong quorum is correct. And therefore that case I suggest you maintain your modification but replacing `>` by `>=`.

Author: Wow 🤯 . The code has defined strong quorum as `> 2/3` for ever. How did we miss this?

Looking back, I see that strong quorum was defined using `>` in the specification docs at least up until [v2](https://docs.google.com/document/d/1Orm58L63SEESM6tDXK9QmWLnnMfF5i-Drj8MB8D3kXw/edit) (and v3 omitted the definition). So this appears to be a concrete change between those earlier documents and the FIP.

If we take a definition of the adversary as strictly <1/3, then it makes sense. But just to confirm: you are proposing that the FIP as it stands is correct with strong quorum as `>= 2/3` everywhere and we should change the code to match.

Author: Actually, @Stebalien changed the code definition from `>` to `>=` in https://github.com/filecoin-project/go-f3/pull/118 when implementing scaled power, but I didn't review that change, or hence notice.

Reviewer 7: Yes, what I meant is that you wrote `return Power(M') + (1-Power(M)) + advPower > ⅔` and I suggest you write `return Power(M') + (1-Power(M)) + advPower >= ⅔` instead. And yes, if the adversarial power is advPower we always can wait for P - advPower units of weight (where P is the total power). Since the adversarial power advPower<1/3 always (because of the LSP82 and DSN88 impossibility results), then P-advPower >= 2P/3 in all cases, so we can wait for >= 2P/3 and this is safe.

In practice for Filecoin this will not matter as the difference is an edge case that will never happen anyways.  (If we wait for >2P/3 then if the adversary is exactly advPower = \lceil P/3 \rceil - 1 then it is possible that we may be waiting for an adversarial node that will never reply, i.e. suppose P = 7 so advPower=2 and we are waiting for >2P/3 or equivalently > 5, then we are waiting for at least 6 votes that we would never get. But this does not matter when P is large enough and even more so when dealing with units of weights and not just number of participants.

#### GossiPBFT pseudocode (main algorithm)

The GossiPBFT algorithm is specified by pseudo-code below. It consists of 3 steps per round (QUALITY/CONVERGE, PREPARE, COMMIT) and an additional step outside the round loop (DECIDE). Message fields that are constant for the instance are omitted for readability. 

Messages are delivered to the participant in some order, whereupon a predicate over the set of received messages is evaluated, depending on the algorithm's current round and step.

> See also the simplified [PlusCal/TLA+ specification](https://github.com/filecoin-project/tla-f3).

Author: Removed

```
GossiPBFT(instance, inputChain, baseChain, committee) → decision, certificate:
// Committee is implicitly used to calculate quora.

01:  round ← 0
03:  proposal ← inputChain  // CHain that the participant locally believes should be decided.
04:  timeout ← 2*Δ
05:  timeout_rebroadcast ← timeout + 1 // Any value larger than timeout, at implementation discretion.
06:  value ← proposal // Used to communicate the voted value to others (proposal or 丄)
07:  evidence ← nil   // Used to communicate evidence for the voted value
08:  C ← {baseChain}
09:  step ← nil
10:  own_msgs ← {} // Set of msgs sent by this participant

11:  while (step != DECIDE)  {
       // QUALTIY
12:    if (round = 0)
13:      BEBroadcast <QUALITY, value>; trigger (timeout)
14:      step ← QUALITY
15:      collect a clean set M of valid QUALITY messages from this instance 
           until HasStrongQuorum(proposal, M) OR timeout expires
17:      C ← C ∪ {prefix : IsPrefix(prefix,proposal) and HasStrongQuorum(prefix,M)}

> 18:      proposal ← longest prefix ∈ C  // At least baseChain, or something longer

Author: Changed the other uses of "heaviest" in F3 to "longest". F3 doesn't know or define chain weight. ("Heaviest" is still the right word when describing EC behaviour, e.g. the heaviest candidate chain)

36:      evidence ← Aggregate(M)  // strong quorum of PREPAREs is evidence
37:    else
38:      value ← 丄  // Vote for not deciding in this round
39:      evidence ← nil

40:    reBroadcast <COMMIT, value, round, evidence>; trigger(timeout)
41:    collect a clean set M of valid COMMIT messages from this round and instance
         until HasStrongQuorumValue(M) OR (timeout expires AND IsStrongQuorum(M))
42:    if (HasStrongQuorumValue(M))
43:      if (StrongQuorumValue(M) ≠ 丄)  // Decide
           // This broadcast sets step = DECIDE which will break the outer loop
44:        evidence ← Aggregate(M)
45:        reBroadcast <DECIDE, StrongQuorumValue(M), evidence>  
46:      else  
           // No participant decided in this round, carry forward existing proposal
47:        evidence ← Aggregate(M)  // Strong quorum of COMMITs for 丄 is evidence
48:    if (∃ m ∈ M: m.value ≠ 丄)  // Value was possibly decided by others
49:      C ← C ∪ {m.value}  // Add to candidate values if not there
50:      proposal ← m.value  // Sway local proposal to possibly decided value
51:      evidence ← m.evidence  // Strong PREPARE quorum is inherited evidence

> 52:    round ← round + 1;

Reviewer 2: ```suggestion
52:    round ← round + 1
```
There are random terminal semicolons sprinkled throughout the pseudocode. Should all be removed for consistency.

Author: Removed the semicolons that are redundantly terminating statements.

```
     // Decide anytime
58:  upon reception of a valid <DECIDE, v, evidence>, if step != DECIDE
59:    reBroadcast <DECIDE, v, evidence>
60:    go to line 56
```

The helper function mayHaveStrongQuorum returns whether, given the already delivered messages and assuming the presence of an equivocating adversary, some correct participant could have observed strong quorum for a value at some round and step. The parameter advPower can be set to 0 in order to consider the possibility of the participant locally observing strong quorum, or ⅓ to consider the possibility of any participant doing so. Note that where only two values are possible, a false return for one of them does not imply that the other has necessarily reached strong quorum (e.g. an exact ⅔ to ⅓ split).

```
61:  mayHaveStrongQuorum(value, round, step, advPower): 
62:    M  ← { m | m.step = step AND m.round = round AND m is valid} // Clean set of messages
63:    M' ← { m | m ∈ M AND m.value = value } 
64:    return Power(M') + (1-Power(M)) + advPower >= ⅔ 
```

> The reBroadcast function broadcasts and remembers a message. It then sets a timeout to rebroadcast a subset of the messages in the current and previous round if the participant cannot terminate the step that it is in. 

Reviewer 2: ```suggestion
The reBroadcast function broadcasts and remembers a message. It then sets a timeout to rebroadcast a subset of the messages in the current and previous round if the participant cannot terminate the step in which it is.
```

103:    deliver(msg)  // Otherwise make the message available to the main algorithm
104:    if (msg' ← shouldJump(round, step) s.t. msg' != nil)
          // The msg' is a CONVERGE for round
105:      round ← msg'.round;
106:      timeout ← 2 * Δ * pow(BackOffExponent, round)
107:      timeout_rebroadcast ← timeout + 1 // Arbitrary increment
108:      if msg'.evidence.step = PREPARE: // Evidence is strong quorum of either PREPAREs of COMMITs for 丄
109:        C ← C ∪ {msg'.value}       // Add to candidate values if not there
110:        proposal ← msg'.value;     // Sway local proposal to possibly decided value
          // Inherit evidence for the value from CONVERGE message 
111:      evidence ← msg'.evidence   
112:      go to line 21  // Start new round>0 by jumping forward

113:  shouldJump(round, step): 
114:    if step = DECIDE:
115:      return nil
116:    if (∃ msg' st. msg'.step = CONVERGE // There must be a CONVERGE for the new round
            AND msg'.round > round          // Round must be greater
            AND ∃ M s.t. M is clean and valid and contains a weak quorum of PREPAREs for msg'.round)
117:      return msg'
118:    return nil
```

> The parameter max_lookahead_rounds may be set at the discretion of an implementation. 

Reviewer 2: ```suggestion
The parameter `max_lookahead_rounds` may be set at the discretion of an implementation. 
```

OR (step = DECIDE                                          | valid DECIDE evidence
      AND (∃ M: IsStrongQuorum(M) AND evidence=Aggregate(M)| is a strong quorum 
        AND ∀ m’ ∈ M: m’.step = COMMIT                     | of COMMIT messages

>         AND ∀ m’,m’’ ∈ M: m’.round = m’’.round             | from the same (any) round as each other

Reviewer 2: Not sure the added "any" contributes to clarity.

        AND ∀ m’ ∈ M: m’.value = value))                   | for the same value 
```

### Evidence verification complexity

> Note that during the validation of each CONVERGE, COMMIT, and DECIDE message, two signatures need to be verified: (1) the signature of the message contents, produced by the sender of the message (first check above) and (2) the aggregate signature in $m.evidence$ ($\texttt{ValidEvidence}(m)$ above). The former can be verified using the sender's public key (stored in the power table). The latter, being an aggregated signature, requires aggregating public keys of all the participants whose signatures it combines. Notice that the evidence of each message can be signed by a different set of participants, which requires aggregating a linear number of public BLS keys (in system size) per evidence verification. However, a participant only needs to verify the evidence once for each *unique* message (in terms of (step, round, value)) received.

Reviewer 2: ```suggestion
Note that during the validation of each CONVERGE, COMMIT, and DECIDE message, two signatures need to be verified: (1) the signature of the message contents, produced by the sender of the message (first check above) and (2) the aggregate signature in $m.evidence$ ($\texttt{ValidEvidence}(m)$ above). The former can be verified using the sender's public key (stored in the power table). The latter, being an aggregated signature, requires aggregating public keys of all the participants whose signatures it combines. Notice that the evidence of each message can be signed by a different set of participants, which requires aggregating a linear number of BLS public keys (in system size) per evidence verification. However, a participant only needs to verify the evidence once for each *unique* message (in terms of (step, round, value)) received.
```

}
```

Certificates for the same instance and decision may vary across participants in which signatures are aggregated to provide the strong quorum.

#### Exchange protocol

A finality certificate is analogous to a full block in the EC block exchange protocol.

> Like with EC block exchange, participants follow **a simple request/response protocol to provide a contiguous sequence of finality certificates**. Note that the sequence of certificates is traversed backward, matching block exchange. 

Reviewer 2: ```suggestion
Like with EC block exchange, participants follow a simple request/response protocol to provide a contiguous sequence of finality certificates. Note that the sequence of certificates is traversed backward, matching block exchange. 
```

No reason to bold.

3. A CID of the power table.
4. The root of the commitements merkle tree.

The epoch is included directly in the `ECTipset` object itself to support proofs of the form "event X happened before/after/at epoch Y", common in cross-chain interactions.

The tipset key is required as this is the main value F3 is finalizing. While off-chain light-clients will use the tipset key, cross-chain bridges likely won't. Thus, the tipset key is _hashed_ before signing (so the full tipset key need not be sent to bridge contracts, saving gas).

> A blake2b hash of the power table is included to aid in syncing the F3 chain. In theory it could be added to the commitment merkle tree and left it out of the root `ECTipset` object, but being able to validate a chain of finality certificates without any additional merkle-proofs is convenient and the an extra 32 bytes is not an unreasonable cost (512 gas on Ethereum) .

Reviewer 2: ```suggestion
A blake2b hash of the power table is included to aid in syncing the F3 chain. In theory, it could be added to the commitment merkle tree and left out of the root `ECTipset` object, but being able to validate a chain of finality certificates without any additional merkle-proofs is convenient and the extra 32 bytes is not an unreasonable cost (512 gas on Ethereum) .
```

#### Wire Format v. Signing Format

> Both the `Payload` and the `ECTIpset`s themselves are encoded them one way for signing and another when sending them over the network.

Reviewer 4: ```suggestion
Both the `Payload` and the `ECTIpset`s themselves are encoded one way for signing and another when sending them over the network.
```

|             | 0 | 1 | 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11 | 12 | 13 | 14 | 15  | 16 | 17 | 18 | 19 | 20 | 21 | 22 | 23 | 24 | 25 | 26 | 27 | 28 | 29 | 30 | 31 |
|-------------|---|---|---|---|---|---|---|---|---|---|----|----|----|----|----|-----|----|----|----|----|----|----|----|----|----|----|----|----|----|----|----|----|
| **00-31**   | G | P | B | F | T | : | f | i | l | e | c  | o  | i  | n  | :  | 0x5 | .  | .  | .  | .  | .  | .  | .  | .  | i  | i  | i  | i  | i  | i  | i  | i  |
| **32-63**   | x | x | x | x | x | x | x | x | x | x | x  | x  | x  | x  | x  | x   | x  | x  | x  | x  | x  | x  | x  | x  | x  | x  | x  | x  | x  | x  | x  | x  |
| **64-95**   | y | y | y | y | y | y | y | y | y | y | y  | y  | y  | y  | y  | y   | y  | y  | y  | y  | y  | y  | y  | y  | y  | y  | y  | y  | y  | y  | y  | y  |
| **96-127**  | z | z | z | z | z | z | z | z | z | z | z  | z  | z  | z  | z  | z   | z  | z  | z  | z  | z  | z  | z  | z  | z  | z  | z  | z  | z  | z  | z  | z  |
| **126-133** | z | z | z | z | z | z |   |   |   |   |    |    |    |    |    |     |    |    |    |    |    |    |    |    |    |    |    |    |    |    |    |    |

> Where `.` is a zero byte, `iii` is the instance (64bit BigEndian integer), `xxx` is the root hash of the per-instance commitmenets tree (from the supplemental data), `yyy` is the root hash of the tipset merkle tree, and `zzz` is the power table CID used to validate the next instance.

Reviewer 4: what is "commitmenets" supposed to be? commitments?

#### Merkle Trees

> Both the tipsets and the commitments inside tipsets are nested into merkle trees to keep proof-sizes (for bridges) small, even if a large instance has many tipsets.

Reviewer 2: ```suggestion
Both the tipsets and the commitments inside tipsets are nested into merkle trees to keep proof sizes small, even if a large instance has many tipsets.
```

The merkle tree is heavily based on the [merkle tree from CometBFT][comet-merkle], but that one:

1. Uses sha256 (not EVM friendly).
2. Requires more complex math on validation. This lets it avoid a few hash operations in some cases, but that likely isn't worth the complexity.
3. Needs additional out-of-band information to be a _true_ vector commitment. Optimization 2 in the CometBFT merkle tree means that a proof for item 7/7 can be treated as a proof of an item 6/6.

> Most other well-known merkle tree formats (verkle trees, etc.) optimize for sparse trees and where this protocol will have _full_ trees, so the complexity isn't worth it.

Reviewer 2: ```suggestion
Most other well-known merkle tree formats (verkle trees, etc.) optimize for sparse trees whereas this protocol will have _full_ trees.
```

* **Safety.** Implementing F3 ensures the safety of finalized outputs during regular or even congested networks against a Byzantine adversary controlling less than ⅓ QAP. For stronger adversaries, F3 provides mitigations to prevent censorship attacks, as outlined above. If deemed necessary, the punishment and recovery from coalitions in the event of an attempted attack on safety can be explored in future FIPs. Note that the (safe) finality time is already significantly improved by F3 compared to the status quo: F3 provides safety of finalized outputs two orders of magnitude faster than the current parameter of 900 epochs during regular network operation.
* **Denial-of-service (DoS).** The implementation of the F3 preserves resistance against DoS attacks currently ensured by Filecoin, thanks to the fully leaderless nature of GossiPBFT and to the use of a VRF to self-assign tickets during the CONVERGE step.

> * **Committees.** This FIP proposes to have all eligible EC participants run all instances of GossiPBFT. While this ensures optimal resilience against a Byzantine adversary, it can render the system unusable if the number of participants grows too large. This introduces an attack vector: if the scalability limit is 100,000 participants, a participant holding as little as 3% of the current QAP can perform a Sybil attack to render the system unusable, with the minimum QAP required per identity. As a result, the implementation should favor the messages of the more powerful participants if the number of participants grows too large. Given that favoring more powerful participants discriminates against the rest, affecting decentralization, amending F3 to use committees in the event of the number of participants exceeding the practical limit might be the topic of a future FIP, as well as the analysis of optimized message aggregation in the presence of self-selected committees.

Reviewer: what's the practical scalability limit today, i.e., if participant # is approaching X, a FIP is necessary to harden the protocol further?
